warehouse/models/intermediate/gtfs/gtfs_quality/int_gtfs_quality__lead_time.sql

```diff
@@ -0,0 +1,249 @@
+         t1.valid_from,
+         t1.next_feed_valid_from,
+         cal_dates.service_id_date,
+         cal_dates.date,
+         cal_dates.exception_type
+    FROM feed_version_history AS t1
+    JOIN dim_calendar_dates AS cal_dates
+      ON t1.feed_key = cal_dates.feed_key
+    -- Since we are comparing feeds with their previous version, omit the initial version of every feed - no comparison is possible
+   WHERE t1.prev_feed_key IS NOT null
+),
+
+-- The self-outer-join, with all of the coalescing, allows us to see:
+------ 1 dates that were REMOVED since the previous version
+------ 2 dates that were ADDED since the previous version
+------ 3 dates that were CHANGED since the previous version
+calendar_dates_version_compare AS (
+  SELECT
+         -- base64_url is same between feed versions
+         COALESCE(cal_dates.base64_url,prev_cal_dates.base64_url) AS base64_url,
+         -- one feed's key is the previous feed's next key
+         COALESCE(cal_dates.feed_key,prev_cal_dates.next_feed_key) AS feed_key,
+         -- one feed's previous key is the previous feed's key
+         COALESCE(cal_dates.prev_feed_key,prev_cal_dates.feed_key) AS prev_feed_key,
+         -- we need to know the next feed's valid_from date, in cases where a trip is removed since the previous feed
+         COALESCE(cal_dates.valid_from,prev_cal_dates.next_feed_valid_from) AS valid_from,
+         cal_dates.exception_type,
+         prev_cal_dates.exception_type AS prev_exception_type,
+         DATE_DIFF(COALESCE(cal_dates.date,prev_cal_dates.date), COALESCE(cal_dates.valid_from,prev_cal_dates.next_feed_valid_from), DAY) AS days_until_date
+    FROM calendar_dates_version_history AS cal_dates
+    FULL OUTER JOIN calendar_dates_version_history AS prev_cal_dates
+      ON cal_dates.prev_feed_key = prev_cal_dates.feed_key
+     AND cal_dates.service_id_date = prev_cal_dates.service_id_date
+),
+
+-- We count each infraction seprately, mostly for QA purposes
+daily_improper_calendar_dates_updates AS (
+  SELECT base64_url,
+         feed_key,
+         valid_from AS date,
+          -- A new calendar_date is being added
+         COUNT(CASE WHEN prev_exception_type IS null THEN 1 END) AS cal_added,
+          -- An existing calendar_date is being removed
+         COUNT(CASE WHEN exception_type IS null THEN 1 END) AS cal_removed,
+          -- An existing calendar_date is changing its exception_type
+         COUNT(CASE WHEN exception_type != prev_exception_type THEN 1 END) AS cal_changed
+    FROM calendar_dates_version_compare
+   WHERE
+        -- Date is between 0 and 7 days from now
+        days_until_date >= 0 AND days_until_date <= 7
+   GROUP BY 1,2,3
+),
+
+feed_update_count AS (
+    SELECT t1.date,
+           t1.feed_key,
+           SUM(t2.trip_added + t2.trip_removed + t2.trip_changed)
+               OVER (
+                   PARTITION BY t2.feed_key
+                   ORDER BY t2.date
+                   ROWS BETWEEN 30 PRECEDING AND CURRENT ROW
```

Comment:
Just to confirm, the check for like September 30 is checking that nothing violated the 7 day rule in the last 30 days? (I was initially confused about why this is 30 and the actual check is 7 day warning)

Reply:
That's correct. The idea is that schedule changes may happen infrequently, and so we want to look at all recent schedule changes to measure adherence to this best practice, rather than saying "did any changes you made on this date break the rule?".